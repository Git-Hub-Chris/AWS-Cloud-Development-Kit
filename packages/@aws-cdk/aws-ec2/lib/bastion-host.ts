import { PolicyStatement } from "@aws-cdk/aws-iam";
import { CfnOutput, Construct } from "@aws-cdk/core";
import { AmazonLinuxGeneration, AmazonLinuxImage, InstanceClass, InstanceSize, InstanceType } from ".";
import { Instance } from "./instance";
import { Port } from "./port";
import { IVpc, SubnetType } from "./vpc";

/**
 * Properties of the bastion host
 */
export interface BastionHostProps {

    /**
     * In which AZ to place the instance within the VPC
     *
     * @default - Random zone.
     */
    readonly availabilityZone?: string;

    /**
     * VPC to launch the instance in.
     */
    readonly vpc: IVpc;

    /**
     * The name of the instance
     *
     * @default - CDK generated name
     */
    readonly instanceName?: string;

    /**
     * Use public subnet instead of private one
     *
     * @default - false
     */
    readonly publicSubnets?: boolean;

    /**
     * Allow classic SSH access via port 22 instead of SSM
     *
     * @default - false
     */
    readonly allowClassicSSH?: boolean;
}

export class BastionHost extends Instance {
    constructor(scope: Construct, id: string, props: BastionHostProps) {
        super(scope, id, {
            vpc: props.vpc,
            availabilityZone: props.availabilityZone,
            instanceName: props.instanceName || 'BastionHost',
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.NANO),
            machineImage: new AmazonLinuxImage({ generation: AmazonLinuxGeneration.AMAZON_LINUX_2 }),
            vpcSubnets: props.publicSubnets ? { subnetType: SubnetType.PUBLIC } : { subnetType: SubnetType.PRIVATE },
        });
        this.addToRolePolicy(new PolicyStatement({
            actions: [
                'ssmmessages:*',
                'ssm:UpdateInstanceInformation',
                'ec2messages:*'
            ],
            resources: ['*'],
        }));
        this.addUserData('yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm');

        if (props.allowClassicSSH) {
            this.connections.allowFromAnyIpv4(Port.tcp(22), 'Allow SSH access from everywhere');
        }

        new CfnOutput(this, 'BastionHostId', {
            description: 'Instance ID of the bastion host. Use this to connect via SSM',
            value: this.instanceId,
        });
    }
}