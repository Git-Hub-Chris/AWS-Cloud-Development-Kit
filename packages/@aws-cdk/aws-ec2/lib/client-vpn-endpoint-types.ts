import { IDependable, IResource } from '@aws-cdk/core';
import { IConnectable } from './connections';

/**
 * A client VPN endpoint
 */
export interface IClientVpnEndpoint extends IResource, IConnectable {
  /**
   * The endpoint ID
   */
  readonly endpointId: string;

  /**
   * Dependable that can be depended upon to force target networks associations
   */
  readonly targetNetworksAssociated: IDependable;
}

/**
 * A Lambda function
 */
export interface IFunction {
  /**
   * The ARN of the function.
   */
  readonly functionArn: string;
}

/**
 * A certificate
 */
export interface ICertificate {
  /**
   * The ARN of the certificate
   */
  readonly certificateArn: string;
}

/**
 * Transport protocol for client VPN
 */
export enum TransportProtocol {
  /** Transmission Control Protocol (TCP) */
  TCP = 'tcp',
  /** User Datagram Protocol (UDP) */
  UDP = 'udp',
}

/**
 * Port for client VPN
 */
export enum VpnPort {
  /** HTTPS */
  HTTPS = 443,
  /** OpenVPN */
  OPENVPN = 1194,
}
