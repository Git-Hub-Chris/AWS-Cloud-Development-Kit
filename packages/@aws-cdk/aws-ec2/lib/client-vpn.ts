import logs = require('@aws-cdk/aws-logs');
import cdk = require('@aws-cdk/core');
import {
  CfnClientVpnAuthorizationRule,
  CfnClientVpnEndpoint,
  CfnClientVpnRoute,
  CfnClientVpnTargetNetworkAssociation
} from './ec2.generated';
import {CIDR_VALIDATION_REGEXES} from "./peer";
import {SecurityGroup} from './security-group';
import {ISubnet, SubnetSelection, Vpc} from './vpc';

interface IClientAuthenticationRequestOptions {
  /**
   * Information about the Active Directory to be used
   */
  readonly activeDirectory?: CfnClientVpnEndpoint.DirectoryServiceAuthenticationRequestProperty;
  /**
   * Information about the authentication certificates to be used
   */
  readonly mutualAuthentication?: CfnClientVpnEndpoint.CertificateAuthenticationRequestProperty;
  /**
   * The type of client authentication to be used.
   * Specify {@link ClientRequestAuthenticationType.CERTIFICATE} to use certificate-based authentication,
   * or {@link ClientRequestAuthenticationType.DIRECTORY_SERVICE} to use Active Directory authentication.
   */
  readonly type: ClientRequestAuthenticationType;
}

/**
 * Authentication method to be used by a Client VPN endpoint
 *
 * @experimental
 */
export class ClientAuthenticationRequest {
  /**
   * Active Directory authentication
   *
   * @param directoryId The ID of the Active Directory to be used for authentication
   */
  public static activeDirectory(directoryId: string): ClientAuthenticationRequest {
    return new ClientAuthenticationRequest({
      activeDirectory: {directoryId},
      type: ClientRequestAuthenticationType.DIRECTORY_SERVICE
    });
  }

  // TODO replace with acm.Certificate?
  /**
   * Certificate-based authentication
   *
   * @param clientRootCertificateChainArn The ARN of the client certificate.
   * The certificate must be signed by a certificate authority (CA)
   * and it must be provisioned in AWS Certificate Manager (ACM).
   */
  public static mutualAuthentication(clientRootCertificateChainArn: string): ClientAuthenticationRequest {
    return new ClientAuthenticationRequest({
      mutualAuthentication: {clientRootCertificateChainArn},
      type: ClientRequestAuthenticationType.CERTIFICATE
    });
  }

  private constructor(public readonly options: IClientAuthenticationRequestOptions) {
  }
}

interface ConnectionLogConfig {
  /**
   * The name of the CloudWatch Logs log group
   */
  readonly cloudwatchLogGroup?: string;
  /**
   * The name of the CloudWatch Logs log stream to which the connection data is published
   */
  readonly cloudwatchLogStream?: string;
  /**
   * Indicates whether connection logging is enabled
   */
  readonly enabled: boolean;
}

/**
 * Client connection logging options for the Client VPN endpoint.
 * If enabled, data about client connections is sent to a Cloudwatch Logs log stream.
 *
 * The following information is logged:
 * * Client connection requests
 * * Client connection results (successful and unsuccessful)
 * * Reasons for unsuccessful client connection requests
 * * Client connection termination time
 *
 * @experimental
 */
export class ConnectionLog {
  /**
   * Log client connection information into a CloudWatch log group
   *
   * @param group CloudWatch log group
   * @param enabled Whether connection logging is enabled
   */
  public static group(group: logs.ILogGroup, enabled = true): ConnectionLog {
    return new ConnectionLog({cloudwatchLogGroup: group.logGroupName, enabled});
  }

  /**
   * Log client connection information into a CloudWatch log stream
   *
   * @param stream CloudWatch log stream
   * @param enabled Whether connection logging is enabled
   */
  public static stream(stream: logs.ILogStream, enabled = true): ConnectionLog {
    return new ConnectionLog({cloudwatchLogStream: stream.logStreamName, enabled});
  }

  private constructor(public readonly options: ConnectionLogConfig) {
  }
}

/**
 *
 * @experimental
 */
export interface ITagSpecificationTag {
  readonly key: string;
  readonly value: string;
}

/**
 *
 * @experimental
 */
export interface ITagSpecification {
  readonly resourceType: TagSpecificationResourceType;
  readonly tags: ITagSpecificationTag[];
}

/**
 *
 * @experimental
 */
export interface ClientVpnEndpointProps {
  /**
   * The VPC in which to create your Client Vpn.
   */
  readonly vpc: Vpc;

  /**
   * Information about the authentication method to be used to authenticate clients
   */
  readonly authenticationOptions: ClientAuthenticationRequest[];

  /**
   * The IPv4 address range, in CIDR notation, from which to assign client IP addresses
   * The address range cannot overlap with the local CIDR of the VPC in which the associated subnet is located,
   * or the routes that you add manually.
   * The address range cannot be changed after the Client VPN endpoint has been created.
   * The CIDR block should be /22 or greater.
   */
  readonly clientCidrBlock: string;
  /**
   * Information about the client connection logging options
   *
   * @default - Creates a new {@link LogGroup}
   */
  readonly connectionLog?: ConnectionLog;

  // TODO replace with acm.Certificate?
  /**
   * The ARN of the server certificate
   */
  readonly serverCertificateArn: string;

  /**
   * A brief description of the Client VPN endpoint
   */
  readonly description?: string;

  /**
   * Information about the DNS servers to be used for DNS resolution.
   * A Client VPN endpoint can have up to two DNS servers.
   * If no DNS server is specified, the DNS address configured on the device is used for the DNS server.
   */
  readonly dnsServers?: string[];

  /**
   * Indicates whether split-tunnel is enabled on the AWS Client VPN endpoint
   *
   * @default false
   * @see [Split-Tunnel on AWS Client VPN Endpoints]{@link https://docs.aws.amazon.com/vpn/latest/clientvpn-admin/split-tunnel-vpn.html}
   */
  readonly splitTunnel?: boolean;

  /**
   * The tags to apply to the Client VPN endpoint during creation
   */
  readonly tagsSpecifications?: ITagSpecification[];

  /**
   * The transport protocol to be used by the VPN session
   *
   * @default ClientVpnEndpointProtocol.UDP
   */
  readonly transportProtocol?: ClientVpnEndpointProtocol;

  readonly routes?: ClientVpnRouteOptions[];
  readonly targetNetworkAssociations?: SubnetSelection[];
  readonly authorizationRules?: ClientVpnAuthorizationRuleOptions[];
}

/**
 *
 * @experimental
 */
export interface IClientVpnEndpoint extends cdk.IResource {
  /**
   * The id of the Client VPN Endpoint
   */
  readonly clientVpnEndpointId: string;
}

/**
 * Define a Client VPN Endpoint
 *
 * A Client VPN endpoint is the resource you create and configure to enable and manage client VPN sessions.
 * It is the destination endpoint at which all client VPN sessions are terminated.
 *
 * @resource AWS::EC2::ClientVpnEndpoint
 * @experimental
 */
export class ClientVpnEndpoint extends cdk.Resource implements IClientVpnEndpoint {

  public readonly clientVpnEndpointId: string;
  public readonly vpc: Vpc;
  public readonly securityGroup: SecurityGroup;

  private routeCount = 0;
  private targetNetworkAssociationCount = 0;
  private authorizationRuleCount = 0;

  constructor(scope: cdk.Construct, id: string, props: ClientVpnEndpointProps) {
    super(scope, id);

    if (!cdk.Token.isUnresolved(props.clientCidrBlock)) {
      const cidrMatch = props.clientCidrBlock.match(CIDR_VALIDATION_REGEXES.ipv4);

      if (!cidrMatch) {
        throw new Error(`Invalid IPv4 CIDR: "${props.clientCidrBlock}"`);
      }

      if (!cidrMatch[2] || Number(cidrMatch[2]) < 22) {
        throw new Error(`CIDR mask should be 22 or greater (got "${cidrMatch[2]}")`);
      }
    }

    if (props.dnsServers && !cdk.Token.isUnresolved(props.dnsServers) && props.dnsServers.length > 2) {
      throw new Error(`A Client VPN endpoint cannot have more than 2 DNS servers, got ${props.dnsServers.length}`);
    }

    const clientVpnEndpoint = new CfnClientVpnEndpoint(this, 'Resource', {
      authenticationOptions: props.authenticationOptions.map(({options}) => options),
      connectionLogOptions: (props.connectionLog ? props.connectionLog : ConnectionLog.group(new logs.LogGroup(scope, 'Log'))).options,
      serverCertificateArn: props.serverCertificateArn,
      clientCidrBlock: props.clientCidrBlock,
      description: props.description,
      splitTunnel: props.splitTunnel,
      tagSpecifications: props.tagsSpecifications,
      dnsServers: props.dnsServers,
      transportProtocol: props.transportProtocol,
    });

    this.clientVpnEndpointId = clientVpnEndpoint.ref;

    this.vpc = props.vpc;
    this.securityGroup = new SecurityGroup(this, 'SecurityGroup', {
      vpc: props.vpc,
      allowAllOutbound: false,
    });

    // TODO Replace with https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/199
    // TODO
    /*new cloudformation.AwsCustomResource(this, 'ApplySecurityGroupsToClientVpnTargetNetwork', {
      onUpdate: {
        service: 'EC2',
        action: 'applySecurityGroupsToClientVpnTargetNetwork',
        parameters: {
          ClientVpnEndpointId: clientVpnEndpoint.ref,
          VpcId: props.vpc.vpcId,
          SecurityGroupIds: [this.securityGroup.securityGroupId],
        },
        physicalResourceId: this.securityGroup.securityGroupId,
      },
    });*/

    if (props.routes) {
      props.routes.map((options) => this.addRoute(options));
    }

    if (props.targetNetworkAssociations) {
      props.targetNetworkAssociations.map((options) =>
        this.addTargetNetworkAssociations(options)
      );
    }

    if (props.authorizationRules) {
      props.authorizationRules.map((options) =>
        this.addAuthorizationRule(options)
      );
    }
  }

  public addRoute(options: ClientVpnRouteOptions): ClientVpnRoute {
    return new ClientVpnRoute(this, `Route-${++this.routeCount}`, {
      clientVpnEndpoint: this,
      ...options,
    });
  }

  /**
   * Each subnet must belong to a different Availability Zone.
   *
   * @param subnetSelection
   */
  public addTargetNetworkAssociations(subnetSelection: SubnetSelection): ClientVpnTargetNetworkAssociation[] {
    // TODO check "Each subnet must belong to a different Availability Zone."
    const {subnets} = this.vpc.selectSubnets(subnetSelection);

    return subnets.map((subnet) =>
      new ClientVpnTargetNetworkAssociation(
        this,
        `TargetNetworkAssociation-${++this.targetNetworkAssociationCount}`, {
          clientVpnEndpoint: this,
          subnet,
        })
    );
  }

  public addAuthorizationRule(options: ClientVpnAuthorizationRuleOptions): ClientVpnAuthorizationRule {
    return new ClientVpnAuthorizationRule(this, `AuthorizationRule-${++this.authorizationRuleCount}`, {
      clientVpnEndpoint: this,
      ...options,
    });
  }
}

/**
 *
 * @experimental
 */
export interface ClientVpnRouteOptions {
  readonly description?: string;
  readonly destinationCidrBlock: string;
  readonly targetSubnet: ISubnet;
}

/**
 *
 * @experimental
 */
export interface ClientVpnRouteProps extends ClientVpnRouteOptions {
  readonly clientVpnEndpoint: IClientVpnEndpoint;
}

/**
 * Define a Client VPN Route
 *
 * @resource AWS::EC2::ClientVpnRoute
 * @experimental
 */
export class ClientVpnRoute extends cdk.Resource {
  constructor(scope: cdk.Construct, id: string, props: ClientVpnRouteProps) {
    super(scope, id);

    new CfnClientVpnRoute(this, 'Resource', {
      clientVpnEndpointId: props.clientVpnEndpoint.clientVpnEndpointId,
      description: props.description,
      destinationCidrBlock: props.destinationCidrBlock,
      targetVpcSubnetId: props.targetSubnet.subnetId,
    });
  }
}

export interface ClientVpnTargetNetworkAssociationOptions {
  readonly subnet: ISubnet;
}

export interface ClientVpnTargetNetworkAssociationProps extends ClientVpnTargetNetworkAssociationOptions {
  readonly clientVpnEndpoint: IClientVpnEndpoint;
}

/**
 * Define a Client VPN Target Network Association
 *
 * @resource AWS::EC2::ClientVpnTargetNetworkAssociation
 * @experimental
 */
export class ClientVpnTargetNetworkAssociation extends cdk.Resource {
  constructor(scope: cdk.Construct, id: string, props: ClientVpnTargetNetworkAssociationProps) {
    super(scope, id);

    new CfnClientVpnTargetNetworkAssociation(this, 'Resource', {
      clientVpnEndpointId: props.clientVpnEndpoint.clientVpnEndpointId,
      subnetId: props.subnet.subnetId,
    });
  }
}

/**
 *
 * @experimental
 */
export interface ClientVpnAuthorizationRuleOptions {
  readonly targetNetworkCidr: string;
  readonly accessGroupId?: string;
  readonly authorizeAllGroups?: boolean;
  readonly description?: string;
}

/**
 *
 * @experimental
 */
export interface ClientVpnAuthorizationRuleProps extends ClientVpnAuthorizationRuleOptions {
  readonly clientVpnEndpoint: IClientVpnEndpoint;
}

/**
 * Define a Client VPN Target Network Association
 *
 * @resource AWS::EC2::ClientVpnAuthorizationRule
 * @experimental
 */
export class ClientVpnAuthorizationRule extends cdk.Resource {
  constructor(scope: cdk.Construct, id: string, props: ClientVpnAuthorizationRuleProps) {
    super(scope, id);

    new CfnClientVpnAuthorizationRule(this, 'Resource', {
      clientVpnEndpointId: props.clientVpnEndpoint.clientVpnEndpointId,
      targetNetworkCidr: props.targetNetworkCidr,
      accessGroupId: props.accessGroupId,
      authorizeAllGroups: props.authorizeAllGroups,
      description: props.description,
    });
  }
}

export enum ClientVpnEndpointProtocol {
  TCP = 'tcp',
  UDP = 'udp',
}

export enum ClientRequestAuthenticationType {
  CERTIFICATE = 'certificate-authentication',
  DIRECTORY_SERVICE = 'directory-service-authentication',
}

export enum TagSpecificationResourceType {
  CLIENT_VPN_ENDPOINT = 'client-vpn-endpoint',
  CUSTOMER_GATEWAY = 'customer-gateway',
  DEDICATED_HOST = 'dedicated-host',
  DHCP_OPTIONS = 'dhcp-options',
  ELASTIC_IP = 'elastic-ip',
  FLEET = 'fleet',
  FPGA_IMAGE = 'fpga-image',
  HOST_RESERVATION = 'host-reservation',
  IMAGE = 'image',
  INSTANCE = 'instance',
  INTERNET_GATEWAY = 'internet-gateway',
  LAUNCH_TEMPLATE = 'launch-template',
  NATGATEWAY = 'natgateway',
  NETWORK_ACL = 'network-acl',
  NETWORK_INTERFACE = 'network-interface',
  RESERVED_INSTANCES = 'reserved-instances',
  ROUTE_TABLE = 'route-table',
  SECURITY_GROUP = 'security-group',
  SNAPSHOT = 'snapshot',
  SPOT_INSTANCES_REQUEST = 'spot-instances-request',
  SUBNET = 'subnet',
  TRAFFIC_MIRROR_FILTER = 'traffic-mirror-filter',
  TRAFFIC_MIRROR_SESSION = 'traffic-mirror-session',
  TRAFFIC_MIRROR_TARGET = 'traffic-mirror-target',
  TRANSIT_GATEWAY = 'transit-gateway',
  TRANSIT_GATEWAY_ATTACHMENT = 'transit-gateway-attachment',
  TRANSIT_GATEWAY_ROUTE_TABLE = 'transit-gateway-route-table',
  VOLUME = 'volume',
  VPC = 'vpc',
  VPC_PEERING_CONNECTION = 'vpc-peering-connection',
  VPN_CONNECTION = 'vpn-connection',
  VPN_GATEWA = 'vpn-gatewa',
}
