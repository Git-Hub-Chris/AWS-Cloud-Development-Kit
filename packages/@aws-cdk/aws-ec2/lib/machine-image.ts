import ssm = require('@aws-cdk/aws-ssm');
import { Construct, Stack, Token } from '@aws-cdk/cdk';
import { UserData } from './user-data';
import { WindowsVersion } from './windows-versions';

/**
 * Interface for classes that can select an appropriate machine image to use
 */
export interface IMachineImageSource {
  /**
   * Return the image to use in the given context
   */
  getImage(scope: Construct): MachineImage;
}

/**
 * Select the latest version of the indicated Windows version
 *
 * The AMI ID is selected using the values published to the SSM parameter store.
 *
 * https://aws.amazon.com/blogs/mt/query-for-the-latest-windows-ami-using-systems-manager-parameter-store/
 */
export class WindowsImage implements IMachineImageSource  {
  constructor(private readonly version: WindowsVersion) {
  }

  /**
   * Return the image to use in the given context
   */
  public getImage(scope: Construct): MachineImage {
    const parameterName = this.imageParameterName(this.version);
    const ami = ssm.StringParameter.valueForStringParameter(scope, parameterName);
    return new MachineImage(ami, OperatingSystem.windows());
  }

  /**
   * Construct the SSM parameter name for the given Windows image
   */
  private imageParameterName(version: WindowsVersion): string {
    return '/aws/service/ami-windows-latest/' + version;
  }
}

/**
 * Amazon Linux image properties
 */
export interface AmazonLinuxImageProps {
  /**
   * What generation of Amazon Linux to use
   *
   * @default AmazonLinux
   */
  readonly generation?: AmazonLinuxGeneration;

  /**
   * What edition of Amazon Linux to use
   *
   * @default Standard
   */
  readonly edition?: AmazonLinuxEdition;

  /**
   * Virtualization type
   *
   * @default HVM
   */
  readonly virtualization?: AmazonLinuxVirt;

  /**
   * What storage backed image to use
   *
   * @default GeneralPurpose
   */
  readonly storage?: AmazonLinuxStorage;
}

/**
 * Selects the latest version of Amazon Linux
 *
 * The AMI ID is selected using the values published to the SSM parameter store.
 */
export class AmazonLinuxImage implements IMachineImageSource {
  private readonly generation: AmazonLinuxGeneration;
  private readonly edition: AmazonLinuxEdition;
  private readonly virtualization: AmazonLinuxVirt;
  private readonly storage: AmazonLinuxStorage;

  constructor(props?: AmazonLinuxImageProps) {
    this.generation = (props && props.generation) || AmazonLinuxGeneration.AMAZON_LINUX;
    this.edition = (props && props.edition) || AmazonLinuxEdition.STANDARD;
    this.virtualization = (props && props.virtualization) || AmazonLinuxVirt.HVM;
    this.storage = (props && props.storage) || AmazonLinuxStorage.GENERAL_PURPOSE;
  }

  /**
   * Return the image to use in the given context
   */
  public getImage(scope: Construct): MachineImage {
    const parts: Array<string|undefined> = [
      this.generation,
      'ami',
      this.edition !== AmazonLinuxEdition.STANDARD ? this.edition : undefined,
      this.virtualization,
      'x86_64', // No 32-bits images vended through this
      this.storage
    ].filter(x => x !== undefined); // Get rid of undefineds

    const parameterName = '/aws/service/ami-amazon-linux-latest/' + parts.join('-');
    const ami = ssm.StringParameter.valueForStringParameter(scope, parameterName);
    return new MachineImage(ami, OperatingSystem.linux());
  }
}

/**
 * What generation of Amazon Linux to use
 */
export enum AmazonLinuxGeneration {
  /**
   * Amazon Linux
   */
  AMAZON_LINUX = 'amzn',

  /**
   * Amazon Linux 2
   */
  AMAZON_LINUX_2 = 'amzn2',
}

/**
 * Amazon Linux edition
 */
export enum AmazonLinuxEdition {
  /**
   * Standard edition
   */
  STANDARD = 'standard',

  /**
   * Minimal edition
   */
  MINIMAL = 'minimal'
}

/**
 * Virtualization type for Amazon Linux
 */
export enum AmazonLinuxVirt {
  /**
   * HVM virtualization (recommended)
   */
  HVM = 'hvm',

  /**
   * PV virtualization
   */
  PV = 'pv'
}

export enum AmazonLinuxStorage {
  /**
   * EBS-backed storage
   */
  EBS = 'ebs',

  /**
   * S3-backed storage
   */
  S3 = 'ebs',

  /**
   * General Purpose-based storage (recommended)
   */
  GENERAL_PURPOSE = 'gp2',
}

/**
 * Construct a Linux machine image from an AMI map
 *
 * Linux images IDs are not published to SSM parameter store yet, so you'll have to
 * manually specify an AMI map.
 */
export class GenericLinuxImage implements IMachineImageSource  {
  constructor(private readonly amiMap: {[region: string]: string}) {
  }

  public getImage(scope: Construct): MachineImage {
    const region = Stack.of(scope).region;
    if (Token.isUnresolved(region)) {
      throw new Error(`Unable to determine AMI from AMI map since stack is region-agnostic`);
    }

    const ami = region !== 'test-region' ? this.amiMap[region] : 'ami-12345';
    if (!ami) {
      throw new Error(`Unable to find AMI in AMI map: no AMI specified for region '${region}'`);
    }

    return new MachineImage(ami, OperatingSystem.linux());
  }
}

/**
 * Representation of a machine to be launched
 *
 * Combines an AMI ID with an OS.
 */
export class MachineImage {
  constructor(public readonly imageId: string, public readonly os: OperatingSystem) {
  }
}

/**
 * The OS type of a particular image
 */
export enum OperatingSystemType {
  LINUX,
  WINDOWS,
}

/**
 * Abstraction of OS features we need to be aware of
 */
export abstract class OperatingSystem {
  /**
   * OS features specialized for Windows
   */
  public static windows(): OperatingSystem {
    class WindowsOS extends OperatingSystem {
      public readonly type = OperatingSystemType.WINDOWS;
      public createDefaultUserData() {
        return UserData.forWindows();
      }
    }
    return new WindowsOS();
  }

  /**
   * OS features specialized for Linux
   */
  public static linux(): OperatingSystem {
    class LinuxOS extends OperatingSystem {
      public readonly type = OperatingSystemType.LINUX;
      public createDefaultUserData() {
        return UserData.forLinux();
      }
    }
    return new LinuxOS();
  }

  public abstract createDefaultUserData(): UserData;

  public abstract get type(): OperatingSystemType;
}
