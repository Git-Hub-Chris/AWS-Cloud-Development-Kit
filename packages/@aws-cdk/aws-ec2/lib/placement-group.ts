import { IResource, Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnPlacementGroup } from './ec2.generated';

export interface IPlacementGroup extends IResource {
  /**
   * The number of partitions. Valid only when Strategy is set to partition.
   */
  partitions?: number;

  /**
   * Determines how this placement group spreads instances
   */
  spreadLevel?: PlacementGroupSpreadLevel;

  /**
   * Determines how this placement group launches instances
   */
  strategy?: PlacementGroupStrategy;

  /**
   * @attribute
   */
  readonly placementGroupName: string;
}

export interface PlacementGroupProps {
  /**
   * The number of partitions. Valid only when Strategy is set to partition.
   */
  readonly partitions?: number;

  /**
   * Determines how this placement group spreads instances
   */
  readonly spreadLevel?: PlacementGroupSpreadLevel;

  /**
   * Determines how this placement group launches instances
   *
   * @default PlacementGroupStrategy.CLUSTER
   */
  readonly strategy?: PlacementGroupStrategy;
}

export enum PlacementGroupSpreadLevel {
  HOST = 'host',
  RACK = 'rack',
}

export enum PlacementGroupStrategy {
  CLUSTER = 'cluster',
  PARTITION = 'partition',
  SPREAD = 'spread',
}

export class PlacementGroup extends Resource implements IPlacementGroup {
  /**
   * The number of partitions. Valid only when Strategy is set to partition.
   */
  partitions?: number;

  /**
   * Determines how this placement group spreads instances
   */
  spreadLevel?: PlacementGroupSpreadLevel;

  /**
   * Determines how this placement group launches instances
   */
  strategy?: PlacementGroupStrategy;

  public readonly placementGroupName: string;

  constructor(scope: Construct, id: string, props?: PlacementGroupProps) {
    super(scope, id);

    this.partitions = props?.partitions;
    this.spreadLevel = props?.spreadLevel;
    this.strategy = props?.strategy;

    if (this.partitions && this.strategy) {
      if (this.strategy !== PlacementGroupStrategy.PARTITION) {
        throw new Error(`PlacementGroup '${id}' can only specify 'partitions' with the 'PARTITION' strategy`);
      }
    } else if (this.partitions && !this.strategy) {
      this.strategy = PlacementGroupStrategy.PARTITION;
    }

    if (this.spreadLevel) {
      if (!this.strategy) {
        this.strategy = PlacementGroupStrategy.SPREAD;
      }
      if (this.strategy !== PlacementGroupStrategy.SPREAD) {
        throw new Error(`PlacementGroup '${id}' can only specify 'spreadLevel' with the 'SPREAD' strategy`);
      }
    }

    const resource = new CfnPlacementGroup(this, 'Resource', {
      partitionCount: this.partitions,
      spreadLevel: this.spreadLevel,
      strategy: this.strategy,
    });

    this.placementGroupName = this.getResourceArnAttribute(resource.attrGroupName, {
      service: 'batch',
      resource: 'compute-environment',
      resourceName: this.physicalName,
    });

  }
}
