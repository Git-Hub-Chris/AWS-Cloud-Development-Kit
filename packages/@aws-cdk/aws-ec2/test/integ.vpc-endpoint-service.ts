import * as cdk from '@aws-cdk/core';
import * as ec2 from '../lib';

const app = new cdk.App();

class VpcEndpointServiceStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const nlbArnNoPrincipals = "arn:aws:elasticloadbalancing:us-east-1:123456789012:loadbalancer/net/Test/9bn6qkf4e9jrw77a";

    new ec2.VpcEndpointService(this, "MyVpcEndpointServiceWithNoPrincipals", {
      networkLoadBalancerArns: [nlbArnNoPrincipals],
      acceptanceRequired: false,
      whitelistedPrincipalIds: []
    });

    const nlbArnWithPrincipals = "arn:aws:elasticloadbalancing:us-east-1:123456789012:loadbalancer/net/Test/1jd81k39sa421ffs";
    const principalArn = "arn:aws:iam::123456789012:root";

    new ec2.VpcEndpointService(this, "MyVpcEndpointServiceWithPrincipals", {
      networkLoadBalancerArns: [nlbArnWithPrincipals],
      acceptanceRequired: false,
      whitelistedPrincipalIds: [principalArn]
    });
  }
}

new VpcEndpointServiceStack(app, 'aws-cdk-ec2-vpc-endpoint-service');
app.synth();
