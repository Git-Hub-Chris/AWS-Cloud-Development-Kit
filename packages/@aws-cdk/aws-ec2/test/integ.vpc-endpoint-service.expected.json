{
  "Resources": {
    "MyVpcEndpointServiceWithNoPrincipals": {
      "Type": "AWS::EC2::VPCEndpointService",
      "Properties": {
        "NetworkLoadBalancerArns": [
          "arn:aws:elasticloadbalancing:us-east-1:123456789012:loadbalancer/net/Test/9bn6qkf4e9jrw77a"
        ],
        "AcceptanceRequired": false
      }
    },
    "MyVpcEndpointServiceWithNoPrincipalsPermissions": {
      "Type": "AWS::EC2::VPCEndpointServicePermissions",
      "Properties": {
        "ServiceId": {
          "Ref": "MyVpcEndpointServiceWithNoPrincipals"
        },
        "AllowedPrincipals": []
      }
    },
    "MyVpcEndpointServiceWithPrincipals": {
      "Type": "AWS::EC2::VPCEndpointService",
      "Properties": {
        "NetworkLoadBalancerArns": [
          "arn:aws:elasticloadbalancing:us-east-1:123456789012:loadbalancer/net/Test/9bn6qkf4e9jrw77a"
        ],
        "AcceptanceRequired": false
      }
    },
    "MyVpcEndpointServiceWithPrincipalsPermissions": {
      "Type": "AWS::EC2::VPCEndpointServicePermissions",
      "Properties": {
        "ServiceId": {
          "Ref": "MyVpcEndpointServiceWithNoPrincipals"
        },
        "AllowedPrincipals": [
          "arn:aws:iam::123456789012:root"
        ]
      }
    }
  }
}